src/logger/auto_logger.c

```diff
@@ -0,0 +1,185 @@
+/*
+ * Race Capture Firmware
+ *
+ * Copyright (C) 2016 Autosport Labs
+ *
+ * This file is part of the Race Capture firmware suite
+ *
+ * This is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *
+ * See the GNU General Public License for more details. You should
+ * have received a copy of the GNU General Public License along with
+ * this code. If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "api.h"
+#include "auto_logger.h"
+#include "convert.h"
+#include "cpp_guard.h"
+#include "dateTime.h"
+#include "gps.h"
+#include "jsmn.h"
+#include "logger.h"
+#include "loggerConfig.h"
+#include "loggerTaskEx.h"
+#include "printk.h"
+#include "serial.h"
+#include <stdbool.h>
+
+#define DEFAULT_START_SPEED_MPH	25
+#define DEFAULT_START_TIME_SEC	5
+#define DEFAULT_STOP_SPEED_MPH	3
+#define DEFAULT_STOP_TIME_SEC	30
+#define LOG_PFX			"[auto_logger] "
+
+static struct {
+        struct auto_logger_config *cfg;
+        tiny_millis_t timestamp_start;
+        tiny_millis_t timestamp_stop;
+} auto_logger_state;
+
+void auto_logger_reset_config(struct auto_logger_config* cfg)
+{
+        cfg->active = true;
+
+        cfg->start.time = DEFAULT_START_TIME_SEC;
+        cfg->start.speed = DEFAULT_START_SPEED_MPH;
+
+        cfg->stop.time = DEFAULT_STOP_TIME_SEC;
+        cfg->stop.speed = DEFAULT_STOP_SPEED_MPH;
+}
+
+static void get_speed_time(struct Serial* serial,
+                           struct auto_logger_speed_time *alst,
+                           const char* name,
+                           const bool more)
+{
+        json_objStartString(serial, name);
+        json_float(serial, "speed", alst->speed, 2, true);
+        json_uint(serial, "time", alst->time, false);
+        json_objEnd(serial, more);
+}
+
+void auto_logger_get_config(struct auto_logger_config* cfg,
+                            struct Serial* serial,
+                            const bool more)
+{
+        json_objStartString(serial, "autoLoggerCfg");
```

Comment:
Thoughts on how would this config structure be extensible for triggering on other sensors in the future?